### PR TITLE
Add clipboard provider configuration

### PR DESCRIPTION
**Pull request by AlfGalf**
This change adds the `clipboard-provider` setting to the `editor` section of configuration.

This option can have values of:
- `none` (use an internal buffer) (on windows only)
- `windows` (use native windows clipboard) (on MacOS only)
- `macos` (use pbcopy/pbpaste) (on neiter of the above)
- `wayland`
- `xclip`
- `xsel`
- `win23yank` (for wsl) (on all targets with "term")
- `termux`
- `tmux`
- `term` (osc codes)
- `custom` (see below for the configuration)

Note for a custom provider the configurations should look like:
```toml
[editor.clipboard-provider.config]
copy = ["tee", "test.txt"]
paste = ["cat", "test.txt"]
primary-copy = ["tee", "test-primary.txt"] # optional
primary-copy = ["cat", "test-primary.txt"] # optional
```

This can be configured at runtime with the usual:
```
set clipboard-provider term
```
Note: I was unable to work out a syntax expression for setting a `custom` provider at runtime. In my opinion this is probably a fine limitation to have but I am curious if there is a correct way I couldn't work out.

This ports over the previous provider selection logic so hopefully the same default behaviour should apply.

I updated the health command to reflect the provider. Note: this required reading the user configurations within the health command which warrants discussion as this seems to not have been done before.

This is my first contribution, I am a C++ developer by profession and a rust hobyist at best so nits and style updates very welcome.

Note: This adds the `nonempty` crate as a new dependency. This is because I wanted a way to fail parsing custom commands if they were empty and didn't want to write custom parsign Serde code. I do not know what the process is for considering new dependencies and am very willing to change this for an alternative solution when presented with a better one! I looked for a `serde` annotation as I thought that was likely but couldn't find one.

### CODE REVIEW

**Comment by AlfGalf** [comment]
Predictably it wont build on machines other than mine 🙃. I'm going to look into cross compiling and clean this up

**Comment by AlfGalf** [comment]
Okay as far as I can tell this is working on Windows, Ubuntu, and MacOS

**Comment by AlfGalf** [comment]
This is a fix for #8826 by the way

**Comment by markstos** [comment]
I would personally put even fewer defaults in the core: target covering 80 to 95% of users. So: Mac, Windows, X11, Wayland, maybe term... and that's it? Once the plugin system is online, I imagine people will be able to provide plugins that support less common options.

**Comment by AlfGalf** [comment]
> I would personally put even fewer defaults in the core: target covering 80 to 95% of users. So: Mac, Windows, X11, Wayland, maybe term... and that's it? Once the plugin system is online, I imagine people will be able to provide plugins that support less common options.

Ive matched the previous modes and default behavior. 
I would be hesitant to remove any supported mode as that may create regressions for users. (Though maybe that comes from too much compiler development)

**Comment by AlfGalf** [comment]
Ping for this PR? 
Do I need to do anything to request reviews?

**Comment by kirawi** [comment]
There's nothing you need to do. It seems like two of the main maintainers are away from Helix right now, so you may have to wait a little.

File: helix-view/Cargo.toml [hunk 1]
```diff
@@ -50,6 +50,7 @@ toml = "0.8"
 log = "~0.4"
 
 parking_lot = "0.12.2"
+nonempty = { version = "0.10.0", features = ["serde", "serialize"] }
```
Comment (the-mikedavis) [inline helix-view/Cargo.toml:53]: We try to keep the set of dependencies reasonably small and I don't think this one pulls its weight. It could be useful to have a wrapper around `Box<[T]>` that has serde implementations in the `helix-stdx` crate for uses like this but we don't need the whole NonEmpty dep

File: helix-view/src/editor.rs [hunk 2]
```diff
@@ -1169,6 +1173,7 @@ impl Editor {
     pub fn refresh_config(&mut self) {
         let config = self.config();
         self.auto_pairs = (&config.auto_pairs).into();
+        self.registers.clipboard_provider = config.clipboard_provider.get_provider();
```
Comment (the-mikedavis) [inline helix-view/src/editor.rs:1176]: I believe this will reset the internal buffers (within NoneProvider for example) when you reload config even if you don't change your clipboard config. I'm not sure there's an easy way to fix this though. We could consider a much larger refactor for clipboard providers to remove state from the clipboard providers (since we store yanks and pastes on the Registers type anyways), and probably remove the clipboard provider trait and just use the enum directly. I'm not sure how large of a change this would take though - I will try out some changes locally and revisit this

File: helix-view/Cargo.toml (hunk 1 above)
Comment (AlfGalf) [reply helix-view/Cargo.toml:53]: Okay great, let me do that

File: helix-view/src/editor.rs (hunk 2 above)
Comment (AlfGalf) [reply helix-view/src/editor.rs:1176]: Yeah, I can think of some gross solutions to this, but this seemed like the least bad option?

File: helix-view/src/register.rs [hunk 3]
```diff
@@ -28,21 +30,19 @@ pub struct Registers {
...
     inner: HashMap<char, Vec<String>>,
-    clipboard_provider: Box<dyn ClipboardProvider>,
+    pub clipboard_provider: Box<dyn ClipboardProvider>,
     pub last_search_register: char,
 }
 
-impl Default for Registers {
-    fn default() -> Self {
+impl Registers {
+    pub fn new(config: Arc<dyn DynAccess<Config>>) -> Self {
```
[the remark this answers is not in the document]
Comment (AlfGalf) [reply helix-view/src/register.rs:38]: Yeah for sure!

**Comment by AlfGalf** [comment]
Sorry for the delay, Ive been traveling, I will try address those comments!

**Comment by AlfGalf** [comment]
Just rebased this onto master

**Comment by AlfGalf** [comment]
Ping?

**Comment by AlfGalf** [comment]
I have made the required changes so I think the "waiting on author" isn't correct

**Comment by the-mikedavis** [comment]
The current tag is waiting-on-review

**Comment by AlfGalf** [comment]
Oh my apologies! I completely misread the log

**Comment by AlfGalf** [comment]
Rebased this to master

**Comment by AlfGalf** [comment]
I have also added documentation for this option.

**Review by the-mikedavis — COMMENTED** [review]
With the current changes introducing a `ClipboardConfig` enum I think it's the right time to refactor this module to use an enum rather than a trait and solve the refactor I was thinking of at the same time (dropping the internal buffer from `FallbackProvider` and switching `set_contents` to take `&self` rather than `&mut self`).

There are a bunch of things about the clipboard module that could use a refactor:

* We should eliminate the `ClipboardProvider` trait and replace it with an enum very similar to what you have in `ClipboardConfig`. There was discussion along these lines when the trait was introduced in https://github.com/helix-editor/helix/pull/310 wondering whether it was worthwhile to use a trait here. I believe the trait makes this module hard to follow and adds some extra complexity we're better off without.
* Currently we use anyhow's Result and Error types here. Instead we should introduce a `ClipboardError` error type using `thiserror` (which we already depend on in helix-view). For `ClipboardProvider::get_contents` for the OSC52 provider we don't implement reading so we should return an appropriate error. In the `Registers` type and module we can match on this error and use the internal buffer instead, eliminating the need for the FallbackProvider to have extra buffers.
* There's a lot of conditional compilation. We need to keep some of it - only building in support for OSC52 and Windows on appropriate platforms and features - but we can eliminate some of it like that the `FallbackProvider` is both a sort of "none" and a "termcode" provider depending on platform. Those can be different `ClipboardProvider` enum members. The conditional compilation also mixes a few states together. Windows now supports OSC52 for example so we should gate the crossterm command's compilation only on `#[cfg(feature = "term")]` where currently it's based on that plus `#[cfg(target_os = "windows")]`. We use conditional compilation to remove any dependencies on `std::process` for WASM targets. I would prefer to just define a dummy `ClipboardProvider` type for WASM that is always a no-op since WASM will not be able to interface with external commands.
* We use the term "copy" which is inconsistent with the rest of the editor: we should use "yank" instead.

I would also drop the use of NonEmpty here. It's pretty small but for the case of specifying external commands we should match what we do for language configuration and use a command and optional (defaulted) args, for example `{ command = "tmux", args = ["save-buffer", "-"] }`.

I have an incomplete draft of what I'm thinking in [this commit](https://github.com/helix-editor/helix/commit/a54dd0fcdefbaa60f9273429462bb15958952c57). The idea for being able to configure any command would be to add a member to that enum mostly identical to `ClipboardConfig::Custom` but reusing `CommandProvider`. Then you could choose between the builtin providers in config by using serde to lowercase/kebab-case their names, like `"wl-clipboard"`, rather than introducing an extra name field.

**Comment by AlfGalf** [comment]
Sorry for the delay on this, been a bit overwhelmed with the day job and _life_.

The latest patch updates this inline with what you suggested above hopefully. I think its in a reasonable state.

The only sins I'm not too comfortable about are the `CommandProvider` and `CommandProviderStatic` structs which I dreamed up to please the ownership system. If there's a better way to achieve this I would be all ears.

**Comment by AlfGalf** [comment]
Sorry for the force pushes, tested on my remote Linux box to check building.

**Comment by AlfGalf** [comment]
There's some windows stuff I don't understand here, that I'd like a look over. Not entirely sure what `execute_winapi` does.
I don't have a windows machine to test on but my ancient Linux box did build for Linux and Windows so there aren't anymore build failures or Linter failures as far as I can tell.

File: helix-view/src/editor.rs [hunk 4]
```diff
@@ -1180,7 +1184,7 @@ impl Editor {
             theme_loader,
             last_theme: None,
             last_selection: None,
-            registers: Registers::default(),
+            registers: Registers::new(conf.clipboard_provider.clone()),
```
Comment (the-mikedavis) [inline helix-view/src/editor.rs:1187]: Now that the clipboard provider is a stateless value we can take advantage of a feature of the `arc_swap` that we use to hold Config and provide the Registers type just that value from config. In `Registers` the `clipboard_provider` field can be a `Box<dyn DynAccess<ClipboardProvider>>` (using `arc_swap::access::DynAccess`) and here in `editor.rs` we can set create this value with

```rust
Box::new(arc_swap::access::Map::new(
    Arc::clone(&config),
    |config: &Config| &config.clipboard_provider,
))
```

this way `registers.rs` only needs to know about this one config value rather than the entire `Config` struct and we can remove the update of the clipboard provider from `refresh_config` in this module. And the `clipboard_provider` field on `Registers` can remain private.

In places where the clipboard provider is read in `register.rs` we would access the field with `&self.clipboard_provider.load()` and functions like `read_from_clipboard` would take a `&ClipboardProvider`

File: helix-view/src/editor.rs [hunk 5]
```diff
@@ -1939,22 +1944,22 @@ impl Editor {
...
-        diagnostics: &'a BTreeMap<Uri, Vec<(lsp::Diagnostic, LanguageServerId)>>,
+    pub fn doc_diagnostics_with_filter<'b>(
+        language_servers: &'b helix_lsp::Registry,
+        diagnostics: &'b BTreeMap<Uri, Vec<(lsp::Diagnostic, LanguageServerId)>>,
         document: &Document,
-        filter: impl Fn(&lsp::Diagnostic, LanguageServerId) -> bool + 'a,
-    ) -> impl Iterator<Item = helix_core::Diagnostic> + 'a {
+        filter: impl Fn(&lsp::Diagnostic, LanguageServerId) -> bool + 'b,
+    ) -> impl Iterator<Item = helix_core::Diagnostic> + 'b {
         let text = document.text().clone();
```
Comment (the-mikedavis) [inline helix-view/src/editor.rs:1963]: I think these lifetime changes are unrelated, right? I think they can be reset

File: helix-view/src/clipboard.rs [hunk 6]
```diff
@@ -1,413 +1,521 @@
...
+        args: Vec<String>,
     }
 
-    impl Default for FallbackProvider {
-        fn default() -> Self {
-            Self::new()
-        }
+    #[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+    #[serde(rename_all = "kebab-case")]
+    pub struct CommandProvider {
```
Comment (the-mikedavis) [inline helix-view/src/clipboard.rs:79]: As you say the StaticCommandProvider + CommandProvider part is not ideal - we can combine these types. You can use a `Cow<'lifetime, BorrowedT>` to represent a type that is either owned or borrowed, so `Command` and `CommandProvider` can be written as

```rust
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct Command<'a> {
    command: Cow<'a, str>,
    #[serde(default)]
    args: Cow<'a, [Cow<'a, str>]>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "kebab-case")]
pub struct CommandProvider<'a> {
    yank: Command<'a>,
    paste: Command<'a>,
    yank_primary: Option<Command<'a>>,
    paste_primary: Option<Command<'a>>,
}
```

For commands that we know at compile time the type will end up with strs and slices borrowed for `'static` and then for custom commands that we need to allocate at runtime we'll use the `Cow::Owned` branch of the enum. For the current `ClipboardConfig` enum this branch would then be `Custom(Box<CommandProvider<'static>>)` (Box because the CommandProvider struct is much larger than the other fields of `ClipboardConfig` - this is a clippy lint).

Then functions like `yank_from_builtin` can take a `provider: CommandProvider<'static>` parameter and `execute_command` can take `cmd: &Command<'_>`. That's the nice part of using a Cow for this type - we can support passing static commands and dynamically created ones with the same signature

Comment (the-mikedavis) [reply helix-view/src/clipboard.rs:79]: Actually it is probably easier to work with these types if we explicitly use the `'static` lifetime:

```rust
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct Command {
    command: Cow<'static, str>,
    #[serde(default)]
    args: Cow<'static, [Cow<'static, str>]>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "kebab-case")]
pub struct CommandProvider {
    yank: Command,
    paste: Command,
    yank_primary: Option<Command>,
    paste_primary: Option<Command>,
}
```

then we can drop the lifetime bounds part anywhere we mention this type, so functions would take `&Command` or `CommandProvider`

File: helix-view/src/clipboard.rs [hunk 7]
```diff
@@ -1,413 +1,521 @@
...
+mod noop {
+    use super::*;
 
-#[cfg(not(target_os = "windows"))]
-pub mod provider {
-    use super::{ClipboardProvider, ClipboardType};
-    use anyhow::Result;
-    use std::borrow::Cow;
+    #[derive(Debug, Clone)]
+    pub enum ClipboardConfig {}
```
Comment (the-mikedavis) [inline helix-view/src/clipboard.rs:45]: minor naming nit but let's keep this type (and the other `ClipboardConfig`s throughout the module) called `ClipboardProvider` - it's configurable now but still a provider

**Comment by AlfGalf** [comment]
I think that's all the comments resolved. Building on Linux and windows just fine also.

**Comment by AlfGalf** [comment]
Thank you for the help with Cow and DynAccess also, Cow at least I had tried to get working and so thats nice to see working

File: Cargo.lock [hunk 8]
```diff
@@ -1332,6 +1332,7 @@ dependencies = [
  "regex-cursor",
  "ropey",
  "rustix",
+ "serde",
```
Comment (the-mikedavis) [inline Cargo.lock:1335]: I believe this change and the one to `helix-stdx/Cargo.toml` can be reset since we don't touch that crate in this PR anymore

File: helix-view/src/register.rs [hunk 9]
```diff
@@ -20,28 +21,25 @@ use crate::{
...
 /// * Primary clipboard (`+`)
-#[derive(Debug)]
 pub struct Registers {
     /// The mapping of register to values.
     /// Values are stored in reverse order when inserted with `Registers::write`.
     /// The order is reversed again in `Registers::read`. This allows us to
     /// efficiently prepend new values in `Registers::push`.
     inner: HashMap<char, Vec<String>>,
-    clipboard_provider: Box<dyn ClipboardProvider>,
+    pub clipboard_provider: Box<dyn DynAccess<ClipboardProvider>>,
```
Comment (the-mikedavis) [inline helix-view/src/register.rs:30]: I think this field can be made private again
```suggestion
    clipboard_provider: Box<dyn DynAccess<ClipboardProvider>>,
```

File: helix-view/src/register.rs [hunk 10]
```diff
@@ -210,17 +213,17 @@ impl Registers {
         }
     }
 
-    pub fn clipboard_provider_name(&self) -> Cow<str> {
-        self.clipboard_provider.name()
+    pub fn clipboard_provider_name(&self) -> Cow<'_, str> {
```
Comment (the-mikedavis) [inline helix-view/src/register.rs:216]: Let's switch this to return a `String` since the `DynAccess` already forces us to own the value and the only caller immediately converts the value to a String anyway

**Comment by AlfGalf** [comment]
Ahh yes, makes sense, resolved

**Comment by sbromberger** [comment]
Just FYI, I tested this using blink.sh (ipad) over mosh and tmux, with provider `termcode` and it works perfectly. Thank you!

**Comment by AlfGalf** [comment]
> Just FYI, I tested this using blink.sh (ipad) over mosh and tmux, with provider `termcode` and it works perfectly. Thank you!

Awesome! I'm really glad to hear it!

**Comment by sbromberger** [comment]
Any chance this could be merged? I've been using it for a week or so and have come to rely on it for clipboard access.

**Review by the-mikedavis — APPROVED** [review]
Looks great, thanks for working on this @AlfGalf! It ended up being a larger refactor than was strictly necessary to make config possible but I really like the way we have `clipboard.rs` laid out now.

**Comment by AlfGalf** [comment]
Thanks for the help with it @the-mikedavis, it's been really fun and educational, I do still have a plan for a hideous asynchronous parsing patch but that will wait until work has quietened down.

**Comment by sbromberger** [comment]
Thank you!!

**Comment by llakala** [comment]
After an update to my flake inputs, the default value here doesn't seem to be properly set. It should be using `wl-copy`, but pasting was simply doing nothing. I've since reverted to an older commit before this, but I can try to get some more replication info.

**Comment by the-mikedavis** [comment]
Ah I see the issue, yank and paste commands are swapped. It's a bit unintuitive because to yank from the clipboard we want the clipboard program to paste and to paste we want it to yank.

**Comment by llakala** [comment]
> Ah I see the issue, yank and paste commands are swapped. It's a bit unintuitive because to yank from the clipboard we want the clipboard program to paste and to paste we want it to yank.

Working great now! Thanks for the quick fix.
